File: 백준/Gold/9663. N－Queen/N－Queen.cc
#include <iostream>
using namespace std;
int n, ret;
bool check[19], ch1[31], ch2[31]; // i + j, i - j

void solve(int idx){
    // 종료
    if(idx == n){
        // 시뮬레이션
        ret++;
        return;
    }

    // 실행, 재귀호출
    for(int j = 0; j < n; j++){ // 행 ↓, 0, 1, 2, ... 
        if(check[j])continue; // 행 열 체크
        if(ch1[idx + j])continue; // '/' 대각선 체크
        if(ch2[idx - j + n])continue; // '\' 대각선 체크 @@@ 

        check[j] = ch1[idx + j] = ch2[idx - j + n] = true;
        // -> 0번째 열에 0번째 행 에다가 놓음

        solve(idx + 1);

        check[j] = ch1[idx + j] = ch2[idx - j + n] = false;
        // -> 원복
    }

}
int main(){
    ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);

    cin >> n;

    solve(0);

    cout << ret << '\n';

    return 0;
}

File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.Buffer;

public class Main {

    static int n, ret;
    static boolean[] ch = new boolean[104];
    static boolean[] ch1 = new boolean[104];
    static boolean[] ch2 = new boolean[104];
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());

        go(0);

        System.out.println(ret);
    }

    static void go(int idx){

        if(idx == n){
            ret++;
            return;
        }

        for(int j = 0; j < n; j++){

            if(ch[j])continue; // idx -> row, j -> col, => one column one queen
            if(ch1[idx + j])continue;
            if(ch2[idx - j + n])continue;

            ch[j] = true; // idx -> row, j -> col, => one column one queen
            ch1[idx + j] = true;
            ch2[idx - j + n] = true;

            go(idx + 1);

            ch[j] = false; // idx -> row, j -> col, => one column one queen
            ch1[idx + j] = false;
            ch2[idx - j + n] = false;
        }
    }
}